contrib/ibmock/verbs.c

```diff
@@ -0,0 +1,1067 @@
+    (void)attr_mask;
+
+    lock();
+    qp_init_attr.qp_context = qp->context;
+    qp_init_attr.send_cq    = qp->send_cq;
+    qp_init_attr.recv_cq    = qp->recv_cq;
+    qp_init_attr.qp_type    = qp->qp_type;
+
+    memcpy(attr, &efa_ib_qp_attr, sizeof(efa_ib_qp_attr));
+    memcpy(init_attr, &qp_init_attr, sizeof(qp_init_attr));
+    unlock();
+    return 0;
+}
+
+int ibv_destroy_qp(struct ibv_qp *qp)
+{
+    struct fake_qp **entry, *fqp = (struct fake_qp*)qp;
+    struct fake_recv_wr *f;
+
+    lock();
+    array_foreach(entry, &fake_qps)
+    {
+        if (fqp == *entry) {
+            array_remove(&fake_qps, entry);
+
+            while (!list_is_empty(&fqp->recv_reqs)) {
+                f = list_first(&fqp->recv_reqs);
+                list_del(&f->list);
+                free(f);
+            }
+
+            free(fqp);
+            unlock();
+            return 0;
+        }
+    }
+
+    printf("ibmock: failed to destroy QP#%d\n", qp->qp_num);
+    unlock();
+    return -1;
+}
+
+/* verbs.h header re/post defines ibv_reg_mr() */
+#undef ibv_reg_mr
+
+struct ibv_mr *ibv_reg_mr_iova2(struct ibv_pd *pd, void *addr, size_t length,
+                                uint64_t iova, unsigned int access)
+{
+    struct fake_pd *fpd = (struct fake_pd*)pd;
+    struct fake_mr *fmr;
+    struct ibv_mr *mr;
+    (void)iova;
+
+    fmr = calloc(1, sizeof(*fmr));
+    if (fmr == NULL) {
+        return NULL;
+    }
+
+    lock();
+    mr          = &fmr->mr;
+    mr->context = pd->context;
+    mr->pd      = pd;
+    mr->addr    = addr;
+    mr->length  = length;
+    mr->handle  = access;
+    mr->rkey = mr->lkey = ++fpd->lkey;
+
+    fmr->fpd = fpd;
+    array_append(&fpd->mrs, &fmr, sizeof(fmr));
+    unlock();
+
+    return mr;
+}
+
+struct ibv_mr *
+ibv_reg_mr(struct ibv_pd *pd, void *addr, size_t length, int access)
+{
+    return ibv_reg_mr_iova2(pd, addr, length, (uint64_t)addr, access);
+}
+
+int ibv_dereg_mr(struct ibv_mr *mr)
+{
+    struct fake_mr **entry, *fmr = (struct fake_mr*)mr;
+
+    lock();
+    array_foreach(entry, &fmr->fpd->mrs)
+    {
+        if (fmr == *entry) {
+            array_remove(&fmr->fpd->mrs, entry);
+            free(fmr);
+            unlock();
+            return 0;
+        }
+    }
+
+    free(fmr);
+    printf("ibmock: failed to deregister MR lkey=%x\n", mr->lkey);
+    unlock();
+    return -1;
+}
+
+int ibv_query_pkey(struct ibv_context *context, uint8_t port_num, int index,
+                   uint16_t *pkey)
+{
+    (void)context;
+    (void)port_num;
+    (void)index;
+
+    if ((port_num != 1) || (index != 0)) {
+        return EINVAL;
+    }
+
+    *pkey = 65535;
+    return 0;
+}
+
+int ibv_query_gid(struct ibv_context *context, uint8_t port_num, int index,
+                  union ibv_gid *gid)
+{
+    struct fake_device *fdevice = (struct fake_device*)context->device;
+    uint8_t addr[16] = {0xfe,        0x80,          0x00,     0x00, 0x00, 0x00,
```

Comment:
a little doc would be useful for this constant

Reply:
fixed